Add hypothesis test for netCDF4 roundtrip

takluyver opened the pull request:
Part of #1846: add a property-based test for reading & writing netCDF4 files.

This is the first time I've played with Hypothesis, but it seems to be working - e.g. I got an error with float16, and the [netCDF docs show](https://www.unidata.ucar.edu/software/netcdf/docs/data_type.html) that 16-bit floats are not a supported data type.

However:

- This currently only tests a dataset with a single variable - it could be extended to multiple variables if that's useful.
- It [looks like](https://www.unidata.ucar.edu/software/netcdf/docs/netcdf_data_set_components.html#Permitted) netCDF4 should support unicode characters, but it failed when I didn't have `max_codepoint=255` in there. I don't know if that's an expected limitation I'm not aware of, or a bug somewhere. But I thought I'd make the test pass for now.

pep8speaks commented:
Hello @takluyver! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`properties/test_netcdf_roundtrip.py`](https://github.com/pydata/xarray/blob/50d242e5a9f2e358c6a8690b96b032a2e867b24c/properties/test_netcdf_roundtrip.py):

> [Line 35:1](https://github.com/pydata/xarray/blob/50d242e5a9f2e358c6a8690b96b032a2e867b24c/properties/test_netcdf_roundtrip.py#L35): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1

dcherian commented:
Thanks @takluyver. LGTM but maybe @Zac-HD can review this and #3285

Zac-HD commented on properties/test_netcdf_roundtrip.py:
```diff
@@ -0,0 +1,49 @@
+compatible_names = st.text(
+    alphabet=st.characters(
+        whitelist_categories=("Ll", "Lu", "Nd"),
+        # It looks like netCDF should allow unicode names, but removing
```
Where is this specified?  It would be useful to provide a link and prehaps explain what's going on with the categories too :smile:

shoyer commented on properties/test_netcdf_roundtrip.py:
```diff
@@ -0,0 +1,49 @@
+        npst.integer_dtypes(),
+        # NetCDF does not support float16
+        # https://www.unidata.ucar.edu/software/netcdf/docs/data_type.html
+        npst.floating_dtypes(sizes=(32, 64)),
```
I think we could add `byte_string_dtypes`, `unicode_string_dtypes`, `datetime64_dtypes` and `timedelta64_dtypes` to this list.

shoyer commented on properties/test_netcdf_roundtrip.py:
```diff
@@ -0,0 +1,49 @@
+    original = xr.Dataset({"data": var})
+    original.to_netcdf(tmp_path / "test.nc")
+
+    roundtripped = xr.open_dataset(tmp_path / "test.nc")
```
It it good practice to use `open_dataset` as a context manager so the file gets explicitly closed.

(Garbage collection should also close the file, but that can't always be relied upon.)

jhamman commented:
@takluyver -

takluyver commented:
Yes, sorry, it dropped off my radar. I've implemented all the changes suggested above.

The datetime64 & timedelta64 cases don't roundtrip completely, so there are test failures. I've added these in a separate commit at the end, so it's easy to cut that back out if you prefer.

Zac-HD commented:
> The datetime64 & timedelta64 cases don't roundtrip completely, so there are test failures. I've added these in a separate commit at the end, so it's easy to cut that back out if you prefer.

The basic problem seems to be that Hypothesis is generating `timedelta64[Y]`, but Xarray is dealing with `timedelta64[ns]`:

```
____________________________ test_netcdf_roundtrip _____________________________
tmp_path = PosixPath('/tmp/pytest-of-vsts/pytest-0/test_netcdf_roundtrip0')
data = data(...), arr = array([293], dtype='timedelta64[Y]')
        ...
        with xr.open_dataset(tmp_path / "test.nc") as roundtripped:
>           xr.testing.assert_identical(original, roundtripped)
E           AssertionError: Left and right Dataset objects are not identical
E           
E           Differing data variables:
E           L   data     (0) timedelta64[ns] -106488 days +01:41:02.290448
E           R   data     (0) timedelta64[ns] -106488 days +01:41:02.290449

properties/test_netcdf_roundtrip.py:51: AssertionError
---------------------------------- Hypothesis ----------------------------------
Falsifying example: test_netcdf_roundtrip(
    tmp_path=PosixPath('/tmp/pytest-of-vsts/pytest-0/test_netcdf_roundtrip0'), 
    data=data(...), 
    arr=array([293], dtype='timedelta64[Y]')
)
```

So either that's a pretty serious bug, or you should specify [the `max_period` of the timedelta dtype strategy](https://hypothesis.readthedocs.io/en/latest/numpy.html#hypothesis.extra.numpy.timedelta64_dtypes).
